peer/peerlist/list.go

```diff
@@ -0,0 +1,588 @@
+// Must be run in a mutex.Lock()
+func (pl *List) removePeerIdentifier(pid peer.Identifier) error {
+	t, err := pl.removePeerIdentifierReferences(pid)
+	if err != nil {
+		// The peer has already been removed
+		return err
+	}
+
+	return pl.transport.ReleasePeer(pid, t)
+}
+
+// removePeerIdentifierReferences will search through the Available and Unavailable Peers
+// for the PeerID and remove it
+// Must be run in a mutex.Lock()
+func (pl *List) removePeerIdentifierReferences(pid peer.Identifier) (*peerThunk, error) {
+	if t := pl.availablePeers[pid.Identifier()]; t != nil {
+		return t, pl.removeFromAvailablePeers(t)
+	}
+
+	if t, ok := pl.unavailablePeers[pid.Identifier()]; ok && t != nil {
+		pl.removeFromUnavailablePeers(t)
+		return t, nil
+	}
+
+	return nil, peer.ErrPeerRemoveNotInList(pid.Identifier())
+}
+
+// removeFromAvailablePeers remove a peer from the Available Peers list the
+// Peer should already be validated as non-nil and in the Available list.
+// Must be run in a mutex.Lock()
+func (pl *List) removeFromAvailablePeers(t *peerThunk) error {
+	delete(pl.availablePeers, t.peer.Identifier())
+	pl.availableChooser.Remove(t, t.Subscriber())
+	t.SetSubscriber(nil)
+	return nil
+}
+
+// removeFromUnavailablePeers remove a peer from the Unavailable Peers list the
+// Peer should already be validated as non-nil and in the Unavailable list.
+// Must be run in a mutex.Lock()
+func (pl *List) removeFromUnavailablePeers(t *peerThunk) {
+	delete(pl.unavailablePeers, t.peer.Identifier())
+}
+
+// Choose selects the next available peer in the peer list
+func (pl *List) Choose(ctx context.Context, req *transport.Request) (peer.Peer, func(error), error) {
+	if err := pl.once.WaitUntilRunning(ctx); err != nil {
+		return nil, nil, pl.newNotRunningError(err)
+	}
+
+	for {
+		p, err := pl.choose(ctx, req)
+		if p != nil {
+			t := p.(*peerThunk)
+			pl.notifyPeerAvailable()
+			t.onStart()
+			return t.peer, t.boundOnFinish, nil
+		}
+		// This is possible only if the available peers chooser is empty at the
+		// time of the call.
+		// We race again each time we add a peer to the available peers.
+		// Since we call Choose outside a lock, it is always possible that it's
+		// empty.
+		if !yarpcerrors.IsUnavailable(err) {
+			return nil, nil, err
+		}
+		if err := pl.waitForPeerAddedEvent(ctx); err != nil {
+			return nil, nil, err
+		}
+	}
+}
+
+func (pl *List) newNotRunningError(err error) error {
+	return yarpcerrors.Newf(yarpcerrors.CodeFailedPrecondition, "%s peer list is not running: %s", pl.name, err.Error())
+}
+
+// IsRunning returns whether the peer list is running.
+func (pl *List) IsRunning() bool {
+	return pl.once.IsRunning()
+}
+
+// choose grabs the next available peer from the PeerRing and returns it,
+// if there are no available peers it returns nil
+func (pl *List) choose(ctx context.Context, req *transport.Request) (peer.StatusPeer, error) {
+	pl.lock.RLock()
+	p, err := pl.availableChooser.Choose(ctx, req)
```

Comment:
do we expect availableChoosers to block until they get an appropriate peer?  We may want to explicitly forbid this behavior since it could cause a deadlock/timeouts (adding a peer requires a lock).